sysconfig/cloud-init: filter MAAS c-i config from ubuntu-seed on grade signed

anonymouse64 opened the pull request:
When the gadget cloud.conf specifies an explicitly allowed datasource, config
for that datasource is permitted to be copied from ubuntu-seed to the run
system for grade signed.

When the gadget cloud.conf specifically does not allow any datasources, then do
not allow any datasources to be copied to the run system for grade signed.

When the gadget cloud.conf just implicitly mentions a datasource, allow that
datasource as well as the statically allowed datasources to be copied to the
run system for grade signed.

When there is no gadget cloud.conf or the cloud.conf does not mention, allow,
or disallow any datasources then only permit config pertaining to specific
datasources that we want to allow and support, specifically MAAS right now, to
be copied to the run system for grade signed.

Dependent PR's:
- [x] https://github.com/snapcore/snapd/pull/10572
- [x] https://github.com/snapcore/snapd/pull/10674
- [x] https://github.com/snapcore/snapd/pull/10588
- [x] https://github.com/snapcore/snapd/pull/10730
- [x] https://github.com/snapcore/snapd/pull/10731
- [x] https://github.com/snapcore/snapd/pull/10732

anonymouse64 commented:
Actually there is a bug in cloud-init handling error state, so this PR has a pre-req of https://github.com/snapcore/snapd/pull/10674 now which hopefully is easy to review and simple to understand.

pedronis commented on sysconfig/cloudinit.go:
```diff
@@ -384,7 +420,35 @@ func configureCloudInit(model *asserts.Model, opts *Options) (err error) {
+			installOpts.AllowedDatasources = nil
+		case gadgetDatasourcesRes.ExplicitlyAllowed != nil:
+			// only allow what the gadget explicitly allowed
+			installOpts.AllowedDatasources = gadgetDatasourcesRes.ExplicitlyAllowed
```
I looked at our original notes and I'm a bit confused, I'm happy to discuss this again but I think we don't want to copy things from seed that are not MAAS, so here we want the intersection (which can be empty) of ExplicitlyAllowed and supportedFilteredDatasources, until we have a use case for something that is not MAAS for this?. Except to make a special case around NoCloud?

anonymouse64 replied:
Hmm I suppose I misunderstood your suggestion when we discussed as I thought it was that what the gadget says always trumps everything else, so if the gadget says `datasource_list: [foo, bar]` then we allow for config from ubuntu-seed for foo and bar (because again this is just for `grade: signed`.

pedronis replied:
but the filtering we do is based on what make sense for MAAS atm, we didn't analyze other things, also for those other datasources they could in theory communicate securely with the source of the data while ubuntu-seed is never secure in that sense

anonymouse64 replied:
Well actually sorry I'm confused now again re-reading your response before pushing the changes. If the gadget mentions a datasource other than MAAS, what should we do with cloud-init config files on ubuntu-seed _that are not specific to a datasource_ like network config. Should we copy it or not? I originally had implemented it such that we still will copy generic config over even if the datasource allowed list was empty, but I can't tell from the doc or your response here if that's what you want to do or not. 

Essentially we have the following cases:
* Gadget mentions MAAS -> great, copy all the config over, filter it for MAAS
* Gadget doesn't mention MAAS -> great, copy all the config over, filter it for MAAS
* Gadget mentions something other than MAAS -> ???
* Gadget explicitly mentions not to allow any datasource -> ???

For the last two cases we have the following options:
1. Install no config files from ubuntu-seed at all, even if they are generic and not specific to a datasource
2. Install _only_ config files that are generic and not specific to a datasource

Or we could do option 1 for the last case where the gadget says it wants no datasources, and option 2 for the case where the gadget says some other datasource.

Config that is not specific to i.e. MAAS can look like this:

```yaml
#cloud-config
network:
  config:
  - id: enp3s0
    mac_address: 52:54:00:b4:9e:25
    mtu: 1500
    name: enp3s0
    subnets:
    - address: 172.16.99.7/24
      dns_nameservers:
      - 172.16.99.1
      type: static
    type: physical
  - address: 172.16.99.1
    type: nameserver
  version: 1
```

pedronis commented on sysconfig/cloudinit.go:
```diff
@@ -384,7 +420,35 @@ func configureCloudInit(model *asserts.Model, opts *Options) (err error) {
+			// config like that?
+
+			installOpts.AllowedDatasources = gadgetDatasourcesRes.Mentioned
+			installOpts.AllowedDatasources = append(installOpts.AllowedDatasources, supportedFilteredDatasources...)
```
for the same reasons I suppose this should just be like default below

pedronis reviewed (commented):
some consideration about the logic. I think we want to mainly support MAAS for now (and NoCloud in case)

anonymouse64 commented on sysconfig/cloudinit.go:
```diff
@@ -301,7 +308,33 @@ func installCloudInitCfgDir(src, targetdir string, opts *cloudInitConfigInstallO
+			src = filteredFile
+		}
+
+		// src may the empty string if we were copying a file that got entirely
```
```suggestion
		// src may be the empty string if we were copying a file that got entirely
```

anonymouse64 commented:
alright this is now implemented as we discussed this morning, it was quite a bit more code and a lot more tests like a lot of tests, but I refactored the tests too so hopefully this is okay to review.

I could split out the cloudDatasourcesInUseForDir to another PR if folks think this has gotten too big, it is over 1k lines now with all the test changes

anonymouse64 commented on sysconfig/cloudinit_test.go:
```diff
@@ -132,41 +138,341 @@ func (s *sysconfigSuite) TestInstallModeCloudInitDisallowedIgnoresOtherOptions(c
+			comment: "no config anywhere, but cloud-init not disabled",
+		},
+		{
+			grade:   "dangerous",
```
I will add more variants here but we did have this existing test: TestInstallModeCloudInitInstallsOntoHostRunModeWithGadgetCloudConfAlsoInstallsUbuntuSeedConfig

but that test can be more easily written as a case inside this test here

anonymouse64 commented on sysconfig/cloudinit.go:
```diff
@@ -443,28 +511,141 @@ func configureCloudInit(model *asserts.Model, opts *Options) (err error) {
+	// so install a restriction file which limits the datasources that can be
+	// used by cloud-init to that which we allowed above, which we know is
+	// non-nil and actually resulted in some files being installed
+	yaml := []byte(fmt.Sprintf(genericCloudRestrictYamlPattern, strings.Join(installOpts.AllowedDatasources, ",")))
```
yeah we weren't testing that scenario but I have tests for it now and we were indeed doing the wrong thing by creating a file with datasources set to be empty. Now we will only write this file if we are filtering files and if there were files installed

anonymouse64 commented on sysconfig/cloudinit.go:
```diff
@@ -443,28 +511,141 @@ func configureCloudInit(model *asserts.Model, opts *Options) (err error) {
+	// non-nil and actually resulted in some files being installed
+	yaml := []byte(fmt.Sprintf(genericCloudRestrictYamlPattern, strings.Join(installOpts.AllowedDatasources, ",")))
+	restrictFile := filepath.Join(ubuntuDataCloudDir(WritableDefaultsDir(opts.TargetRootDir)), "cloud.cfg.d/99_snapd_datasource.yaml")
+	return ioutil.WriteFile(restrictFile, yaml, 0644)
```
yes there will end up being actually two files which restrict cloud-init, 99_snapd_datasource.cfg here if we installed and filtered some files from ubuntu-seed, and then again later on when snapd runs the first time with zzzz_snapd.cfg

anonymouse64 commented on sysconfig/cloudinit.go:
```diff
@@ -443,28 +511,141 @@ func configureCloudInit(model *asserts.Model, opts *Options) (err error) {
+	// used by cloud-init to that which we allowed above, which we know is
+	// non-nil and actually resulted in some files being installed
+	yaml := []byte(fmt.Sprintf(genericCloudRestrictYamlPattern, strings.Join(installOpts.AllowedDatasources, ",")))
+	restrictFile := filepath.Join(ubuntuDataCloudDir(WritableDefaultsDir(opts.TargetRootDir)), "cloud.cfg.d/99_snapd_datasource.yaml")
```
yes it should be

anonymouse64 commented on tests/nested/manual/core20-cloud-init-maas-signed-seed-data/task.yaml:
```diff
@@ -0,0 +1,187 @@
+  SNAKEOIL_CERT="$PWD/$KEY_NAME.pem"
+
+  snap download --basename=pc --channel="20/edge" pc
+  unsquashfs -d pc-gadget pc.snap
```
probably not a bad idea, but I will push the rest of the changes before working on that as the spread test is finnicky to iterate on being a nested test and all

pedronis replied:
can also be a follow up as this PR is already big, but yes I think it would be a good idea to have it

anonymouse64 replied:
ack I think doing this in a followup sounds good

anonymouse64 replied:
this is now done in https://github.com/snapcore/snapd/pull/10807

anonymouse64 commented on tests/nested/manual/core20-cloud-init-maas-signed-seed-data/task.yaml:
```diff
@@ -0,0 +1,187 @@
+  tests.nested exec "test -f /etc/cloud/cloud.cfg.d/90_50-cloudconfig-maas-datasource.cfg"
+  tests.nested exec "test -f /etc/cloud/cloud.cfg.d/90_50-cloudconfig-maas-reporting.cfg"
+  tests.nested exec "test -f /etc/cloud/cloud.cfg.d/90_50-curtin-networking.cfg"
+
```
oh well I guess I just added a check for the presence, I will expand the check for the contents when I add the cloud.conf in the gadget variant

anonymouse64 replied:
This is also done in https://github.com/snapcore/snapd/pull/10807

pedronis reviewed (approved):
thanks

mvo5 reviewed (approved):
Thanks for working on this! Looks fine, I had some minor comments on things like var names or placing but decided to not include it as it's more a personal taste thing.
